pkgs/crane-utils/src/bin/crane-resolve-workspace-inheritance.rs

```diff
@@ -105,90 +96,300 @@ fn merge(cargo_toml: &mut toml::Value, root: &toml::Value) {
         ("dev-dependencies", "dependencies"),
         ("build-dependencies", "dependencies"),
     ] {
-        if let (Some(toml::Value::Table(p)), Some(toml::Value::Table(wp))) =
-            (t.get_mut(key), w.get(ws_key))
-        {
-            merge_tables(p, wp);
+        if let Some((cargo_toml, root)) = cargo_toml.get_mut(key).zip(w.get(ws_key)) {
+            try_merge_cargo_tables(cargo_toml, root);
         };
 
-        if let Some(toml::Value::Table(targets)) = t.get_mut("target") {
-            for (_, tp) in targets {
-                if let (Some(toml::Value::Table(p)), Some(toml::Value::Table(wp))) =
-                    (tp.get_mut(key), w.get(ws_key))
-                {
-                    merge_tables(p, wp);
-                };
+        if let Some(targets) = cargo_toml.get_mut("target").and_then(try_as_table_like_mut) {
+            for (_, tp) in targets.iter_mut() {
+                if let Some((cargo_toml, root)) = tp.get_mut(key).zip(w.get(ws_key)) {
+                    try_merge_cargo_tables(cargo_toml, root);
+                }
             }
         }
     }
+
+    // cleanup stray spaces left by merging InlineTable values into Tables
+    clear_value_formatting(cargo_toml.as_item_mut());
+}
+
+/// Return a [`toml_edit::TableLike`] representation of the [`Item`] (if any)
+fn try_as_table_like(item: &Item) -> Option<&dyn toml_edit::TableLike> {
+    match item {
+        Item::Table(w) => Some(w),
+        Item::Value(toml_edit::Value::InlineTable(w)) => Some(w),
+        _ => None,
+    }
 }
 
-fn merge_tables(cargo_toml: &mut Table, root: &Table) {
+/// Return a mutable [`toml_edit::TableLike`] representation of the [`Item`] (if any)
+fn try_as_table_like_mut(item: &mut Item) -> Option<&mut dyn toml_edit::TableLike> {
+    match item {
+        Item::Table(w) => Some(w),
+        Item::Value(toml_edit::Value::InlineTable(w)) => Some(w),
+        _ => None,
+    }
+}
+
+/// Merge the specified `cargo_toml` and workspace `root` if both are tables
+fn try_merge_cargo_tables(cargo_toml: &mut Item, root: &Item) {
+    match (cargo_toml, root) {
```

Comment:
Would it be simpler to use `try_as_table_like{,mut}` instead of matching on every combination of the variants? Or is the goal to avoid the dynamic dispatch?

Reply:
I updated the comment to clarify: the way we are performing the merge, the "surrounding decoration" is only copied over values that need to be added.   Since this top-level comment is on a key "foo" which is in common with cargo_toml, the keys are merged and the comment is taken from the "cargo_toml" side.

I also removed the `fmt` module in a separate commit, to show how the resulting `smoke_test` truth has to change.  Do you think this is OK as-is (string comparison), or should I try using the `toml` crate as a dev-dependency to compare just TOML content, to restore more sane spacing?